system/core/tags.php

```diff
@@ -0,0 +1,184 @@
+<?php
+	// TODO:
+	// - add styling for default theme
+	// - implement error notification if new tag exists
+	// - implement tag management page: edit, translate, remove or merge tags
+	// - implement tagcontainer pagetype
+	
+	function tagList($pageListNode, $lang) {
+		global $hyphaXml;
+		
+		// Build list of active tags for given page
+		foreach($pageListNode->getElementsByTagName('tag') as $tag) $activeTags[] = $tag->getAttribute('id');
+		
+		// Divide list of all tags stored in hypha.xml in a selected tag list and a deselected tag list
+		foreach($hyphaXml->getElementsByTagName('tag') as $tag) {
```

Comment:
This also finds *all* tag elements, which I think is not intended here? E.g. it finds the tag definitions, but also the tag references inside the page list. In practice the latter do not have any `language` children elements, so this doesn't break, but it is at least inefficient and prone to breakage in the future.

I would suggest using Xpath, e.g. something like `$hyphaXml->findXPath("/tags/tag")` (not tested, but I think `$hyphaXml` should have a `findXPath()` method, and I'm just guessing about the XML structure that you need to match with this Xpath expression).

Reply:
I wonder if it would be a good idea to let the xpath query lookup the tags by language (and perhaps id)? That might make the code simpler.

Reply:
Laurens implemented both suggestions.